### PR TITLE
sc apps: Dashboard for visualizing how spread-out pods are across nodes

### LINKED ISSUE #1345: Make it possible to show zone/node spread for pods in grafana
**Is your feature request related to a problem? Please describe.**
Users might want to be able to easily see how well their pods are spread across zones/nodes.

**Describe the solution you'd like**
I have assumed that the grafana dashboard would be for a specific deployment/statefulset. But you could perhaps in addition/instead do it for all deployments/statefulsets in a namespace/cluster.

To visualize node-spread it should be enough to use the metric `kube_pod_info`, which includes info about which node a pod is running on. To visualize zone spread I don't think we have any metric that shows which zone a pod or a node is in. But kube-state-metrics can expose node labels and we usually have the labels `topology.kubernetes.io/zone` and `failure-domain.beta.kubernetes.io/zone` to show which zone a node is in. So adding --metric-labels-allowlist=nodes=[topology.kubernetes.io/zone] should be enough to get the missing info. Then we can combine that with the info about which node a pod is running on.

Not sure how to visualize this in a good way in grafana.

**Describe alternatives you've considered**

**Additional context**
[Documentation with `kube-state-metrics` arguments.](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/cli-arguments.md)

**Definition of done:**
There is a grafana dashboard that can show how a deployment/statefulset is spread across zones/nodes.

### PR DESCRIPTION
**Pull request by Zash**
> [!warning]
> **This is public repository, ensure not to disclose:**
> - [x] personal data beyond what is necessary for interacting with this pull request
> - [x] business confidential information, such as customer names

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [x] kind/feature       
- [ ] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change  
- [ ] kind/dev-change    
- [ ] kind/security      
- [ ] [kind/adr]()       

### What does this PR do / why do we need this PR?

This dashboard lets you explore how well Pods are spread out across either Nodes or Zones. 

- Fixes #1345

#### Screenshots

image

#### Checklist

- [x] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [ ] The change requires no migration steps
  - [ ] The change requires migration steps
- Metrics checks:
  - [x] The metrics are still exposed and present in Grafana after the change
  - [x] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [x] The logs do not show any errors after the change
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packages in the `NetworkPolicy Dashboard`
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Audit checks:
  - [x] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy

### CODE REVIEW

File: WIP-CHANGELOG.md [hunk 1]
```diff
@@ -40,6 +40,7 @@
   - Add pattern matching for node and disk
 - Added support for configuring multiple kube-apiservers for blackbox exporter in sc
 - `./bin/ck8s update-ips` now adds the IPs and port for swift endpoint
+- Added dashboard that showing how spread out pods are across nodes
```
Comment (OlleLarsson) [inline WIP-CHANGELOG.md:43]:
```suggestion
- Added dashboard that shows how spread out pods are across nodes
```

**Review by davidumea — COMMENTED** [review]
If I understood the issue correctly, the dashboard should also show zone skew, e.g. by using label `topology.kubernetes.io/zone`. Problem is that **most** of the infra providers that we use at least have all nodes in the same zone. (except e.g. elastx)

**Comment by Zash** [comment]
@davidumea 
> Problem is that **most** of the infra providers that we use at least have all nodes in the same zone.

My intent was to do add a second dashboard for zones in a second PR. Are you saying it can be skipped?

**Comment by davidumea** [comment]
> My intent was to do add a second dashboard for zones in a second PR.

Okay, I think it would be nice to have it in the same dashboard, that way the user can find all information about how a deployment is spread in one place.

> Are you saying it can be skipped?

No I don't think it should be skipped :slightly_smiling_face:. I meant that maybe we want to do something extra here, possibly sort out/not show the data if there only is one zone in the environment. If more than one zone exists then it should be obvious if multiple replicas of a deployment are in the same zone.

**Comment by Zash** [comment]
Pushing an updated version where you can pick between node or zone skew.

**Comment by Zash** [comment]
image

**Review by davidumea — APPROVED** [review]
This looks really good :+1:

**Comment by viktor-f** [comment]
The "pods per node" graph, is that stacked? I think I would prefer if that was not stacked. I think that it might then be easier to see how many pods are on each node.

**Comment by viktor-f** [comment]
> > The "pods per node" graph, is that stacked? I think I would prefer if that was not stacked. I think that it might then be easier to see how many pods are on each node.
> 
> Not stacked currently, so lines with the same number of pods overlap.
> 
If they are not stacked, then something seems wrong with the last screenshot you sent. For rook-ceph it is reporting a skew of 2, but looking at the graph I think it should be 6-3=3. Am I missing something?

> > This is shown per deployment/statefulset, right?
>
>  Currently per namespace.
Didn't you have it working for deployment/statefulset earlier? This could still be useful, but it would be much better to see the skew per deployment/statefulset since that is what kubernetes should be balancing.

**Comment by Zash** [comment]
> If they are not stacked, then something seems wrong with the last screenshot you sent. For rook-ceph it is reporting a skew of 2, but looking at the graph I think it should be 6-3=3. Am I missing something?

That's shown per node but the view is set to zone granularity, so it's the difference between 8 pods in zone 'a' and 6 pods in zone 'b'. 

> Didn't you have it working for deployment/statefulset earlier? This could still be useful, but it would be much better to see the skew per deployment/statefulset since that is what kubernetes should be balancing.

Sorta, but I had trouble keeping all the labels trough joins with variable granularity. Trying again this morning, I made some progress on that.

**Comment by Zash** [comment]
After talking with @viktor-f I'll try to rearrange the per-Namespace graphs to loop over ReplicaSet / StatefulSet instead

**Comment by Zash** [comment]
Now with per-ReplicaSet/StatefulSet calculations:

image

File: helmfile/upstream/prometheus-community/kube-prometheus-stack/charts/kube-state-metrics/values.yaml [hunk 2]
```diff
@@ -233,7 +233,8 @@ metricDenylist: []
 # label keys you would like to allow for them (Example: '=namespaces=[k8s-label-1,k8s-label-n,...],pods=[app],...)'.
 # A single '*' can be provided per resource instead to allow any labels, but that has
 # severe performance implications (Example: '=pods=[*]').
-metricLabelsAllowlist: []
+metricLabelsAllowlist:
+  - nodes=[topology.kubernetes.io/zone]
```
Comment (OlleLarsson) [inline helmfile/upstream/prometheus-community/kube-prometheus-stack/charts/kube-state-metrics/values.yaml:237]: Isn't it possible to set this in the kube-prometheus-stack values file, e.g. [here](https://github.com/elastisys/compliantkubernetes-apps/blob/main/helmfile/values/kube-prometheus-stack-wc.yaml.gotmpl#L23)?

Comment (Zash) [reply helmfile/upstream/prometheus-community/kube-prometheus-stack/charts/kube-state-metrics/values.yaml:237]: Good question. Still new to all these layers of indirection 😄 

Something along these lines perhaps?

``` diff
@@ -21,6 +21,13 @@ prometheusOperator:
     resources: {{- toYaml .Values.prometheusOperator.prometheusConfigReloader.resources | nindent 6 }}
 
 kube-state-metrics:
+
+  # if zones enabled
+  {{- if true }}
+  metricLabelsAllowlist:
+  - nodes=[topology.kubernetes.io/zone]
+  {{- end }}
+
   resources: {{- toYaml .Values.kubeStateMetrics.resources | nindent 4 }}
   {{- if .Values.kubeStateMetrics.clusterAPIMetrics.enabled }}
   rbac:
```

Is there an existing settings for whether zones are used at all yet?

Comment (OlleLarsson) [reply helmfile/upstream/prometheus-community/kube-prometheus-stack/charts/kube-state-metrics/values.yaml:237]: > Is there an existing settings for whether zones are used at all yet?

No, not that i'm aware of

Comment (Zash) [reply helmfile/upstream/prometheus-community/kube-prometheus-stack/charts/kube-state-metrics/values.yaml:237]: Hm, does it even need to be conditional?

Comment (OlleLarsson) [reply helmfile/upstream/prometheus-community/kube-prometheus-stack/charts/kube-state-metrics/values.yaml]: Not entirely sure, you could see what happens if the label isn't set on the nodes.

Comment (Zash) [reply helmfile/upstream/prometheus-community/kube-prometheus-stack/charts/kube-state-metrics/values.yaml]: If it's in the allow-list but no such labels are set it just doesn't show up in the metrics data, so I don't think there'll be any storage overhead for clusters without it. Unless there's some significant optimization for the case of an _empty_ allow-list, I doubt there's much of a performance impact either.

So not adding conditional or new settings for now then.

**Comment by Zash** [comment]
Updated first comment with new screenshot. Dropped multi-select for namespaces because it caused wrong results. Dropped the overview for similar reason.

**Review by viktor-f — APPROVED** [review]
LGTM, we can see if we can figure out an overview later.

**Pull request merged**